### PR TITLE
PRC-56 : Additional date of birth validation

### PR DESCRIPTION
**Pull request by james-baker-moj**
Ensure day, month and year and in valid ranges and that the date of birth is not in the future.

### CODE REVIEW

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts [hunk 1]
```diff
@@ -2,39 +2,61 @@ import z from 'zod'
...
 export const createContactEnterDobSchema = () => async () => {
   return createSchema({
     isDobKnown: z.string({ message: SELECT_IS_DOB_KNOWN_MESSAGE }).transform(value => value === 'true'),
-    day: z.coerce.number({ message: DAY_TYPE_MESSAGE }).optional(),
-    month: z.coerce.number({ message: MONTH_TYPE_MESSAGE }).optional(),
-    year: z.coerce.number({ message: YEAR_TYPE_MESSAGE }).optional(),
+    day: z.union(
+      [
+        z.literal(''),
+        z.coerce.number({ message: DAY_TYPE_MESSAGE }).min(1, DAY_TYPE_MESSAGE).max(31, DAY_TYPE_MESSAGE).optional(),
```
Comment (prabash-moj) [inline server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts:16]: formatting to make it into new lines
`.coerce.number({ message: DAY_TYPE_MESSAGE })

.min(1, DAY_TYPE_MESSAGE)

.max(31, DAY_TYPE_MESSAGE)

.optional(),`

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts [hunk 2]
```diff
@@ -2,39 +2,61 @@ import z from 'zod'
...
+        z.coerce
+          .number({ message: MONTH_TYPE_MESSAGE })
+          .min(1, MONTH_TYPE_MESSAGE)
+          .max(12, MONTH_TYPE_MESSAGE)
+          .optional(),
+      ],
+      { errorMap: () => ({ message: MONTH_TYPE_MESSAGE }) },
+    ),
+    year: z.union(
+      [z.literal(''), z.coerce.number({ message: YEAR_TYPE_MESSAGE }).min(1900, YEAR_TYPE_MESSAGE).optional()],
```
Comment (prabash-moj) [inline server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts:32]: same here

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.test.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+      const result = await doValidate(form)
+
+      // Then
+      expect(result.success).toStrictEqual(true)
+      expect(result.data).toStrictEqual({
+        isDobKnown: false,
+      })
+    })
+    it('should handle empty/invalid dob fields if dob is not known', async () => {
+      // Given - we can select yes and enter day, month and year and then change to no so they may be populated in the form
```
Comment (prabash-moj) [inline server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.test.ts:208]: is this comment necessary here, if it is may be update the test method description to reflect this ?

**Review by prabash-moj — APPROVED** [review]

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts (hunk 1 above)
Comment (james-baker-moj) [reply server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts:16]: I did but then the formatter put it back like this 🤷

**Review by prabash-moj — APPROVED** [review]

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**